File: python/stemel.py
import re

def stemel_rest(step_size):
  return {"rest":step_size}

def replace_rests(pitch, duration, sustain, obj):
  new_duration = []
  for step in duration:
    if type(step)==type((1,)):
      new_dur = ()
      for dur in step:
        if type(dur) == type({}):
          new_dur = (*new_dur, obj)
        else:
          new_dur = (*new_dur, dur)
      new_duration.append(new_dur)
    else:
      if type(step) == type({}):
        new_duration.append(obj)
      else:
        new_duration.append(step)
  return (pitch, new_duration, sustain)

def separate(notes):
  """
  separate a score of notes into a tuple of
  frequencies, sustains and durations.
  """
  frequencies = []
  durations = []
  sustains = []
  if len(notes)==0:
    return ([],[])
  length = len(max(notes, key = lambda x: len(x)))
  counter = 0
  for i in range(0, length):
    freq_buffer = ()
    dur_buffer = ()
    sus_buffer = ()
    for voice_line in notes:
      if len(voice_line)>0:
        freq_buffer = (*freq_buffer, voice_line[counter % len(voice_line)]['frequency'])
        sus_buffer = (*sus_buffer, voice_line[counter % len(voice_line)]['sustain'])
        dur_buffer = (*dur_buffer, voice_line[counter % len(voice_line)]['duration'])
    new_duration_buff = ()
    for duration in dur_buffer:
      new_duration_buff = (*new_duration_buff, duration)
    new_frequency_buff = ()
    for frequency in freq_buffer:
      if (frequency > -1):
        new_frequency_buff = (*new_frequency_buff, frequency)
    if len(new_frequency_buff)==0:
      #add a placeholder
      new_frequency_buff=(0,)
    if len(new_frequency_buff)>1:
      frequencies.append(new_frequency_buff)
    else:
      #replace single tuple with single value
      frequencies.append(new_frequency_buff[0])
    if len(new_duration_buff)>1:
      durations.append(new_duration_buff)
    elif len(new_duration_buff) == 1:
      durations.append(new_duration_buff[0])
    else:
      durations.append(0)
    if len(sus_buffer)>1:
      sustains.append(sus_buffer)
    else:
      sustains.append(sus_buffer[0])
    counter += 1
  return (frequencies, durations, sustains)

def make_pattern(score, step_size=0.25):
  """
  Take a score inputted as a line of text and a step size (0.25 default)
  and return a tuple of frequencies, sustains and durations
  """
  polyvals=[]
  vals = []
  polyvals.append(vals)
  oct = 0
  amplitude = 1.0
  for note in re.split(r'\s+', score):
    note = note.strip()
    note = note.lower()
    if re.search(r'^-+?', note):
      # carry previous note one more step
      for i in note:
        lookback_counter = len(vals)-1
        while (lookback_counter > 0) and (vals[lookback_counter]['frequency']<0):
          lookback_counter -=1
        if lookback_counter>=0:
          vals[lookback_counter]["sustain"]+=step_size
          vals[lookback_counter]["duration"]+=step_size
          print(vals[lookback_counter]["duration"])
    elif re.search(r'^\*+?$', note):
      # rest
      for i in note:
         vals.append({'frequency':-1,'duration':stemel_rest(step_size),'sustain':step_size})
    elif re.search(r'/',note):
      # new track
      vals = []
      polyvals.append(vals)
    elif re.search(r'^>+?', note):
      # octave up
      for i in note:
        oct +=1
    elif re.search(r'^<+?', note):
      # octave down
      for i in note:
        oct -=1
      if oct < 0:
        oct = 0
    elif re.search(r'^[\d | \.]+?$', note):
      # note
      vals.append({'frequency':(float(note)+(oct*12)),'duration':step_size,'sustain':step_size})
  return separate(polyvals)

File: ruby/lib/stemel.rb
def make_notes(pattern_string)
  polyvals=[]
  vals = []
  polyvals << vals
  oct = 3
  amplitude = 1.0
  pattern_string.split(" ").each do |note|
    note.strip!
    note.downcase!
    case note
    when /^-+?/
      note.size().times do
        lookback_counter = vals.size()-1
        while ((lookback_counter > 0) && (vals[lookback_counter][:frequency]<0))
          lookback_counter -=1
        end
        if lookback_counter>=0
          vals[lookback_counter][:sustain]+=1
        end
        vals << {:frequency => -1, :duration=> {"rest":1}, :sustain => 1}
      end
    when /^\*+?$/
      # rest
      note.size().times do
        vals << {:frequency => -1, :duration=> {"rest":1}, :sustain => 1}
      end
    when '/'
      vals = []
      polyvals << vals
    when /^>+?/
      note.size().times do
        oct +=1
      end
    when /^<+?/
      note.size().times do
        oct -=1
      end
      if oct < 0
        oct = 0
      end
    when /^[\d | \.]+?$/
        vals << { :frequency => note.to_f+(oct*12), :duration=> 1, :sustain => 1}
    else
      puts "? " + note
    end
  end
  polyvals
end

def separate(pattern, step_size)
  durations = []
  sustains = []
  pitches = []
  pattern.each do |line|
    dur_buff =[]
    sus_buff =[]
    pit_buff =[]
    line.each do |note|
      duration = note[:duration]
      if duration.class == Hash
        duration[:rest]=step_size
        dur_buff << duration
      else
        dur_buff << duration*step_size
      end
      sus_buff << note[:sustain]*step_size
      pit_buff << note[:frequency]
    end
    durations << dur_buff
    sustains << sus_buff
    pitches << pit_buff
  end
  {:pitches => pitches, :durations => durations, :sustains => sustains}
end

def make_pattern(score, step_size)
  pattern = []
  notes = make_notes(score)
  return if notes.size()==0
  length = notes.max{ |a,b| a.size() <=> b.size()}.size()
  counter = 0
  length.times do
    seq_buffer = []
    notes.each do |voice_line|
      if (voice_line.size()>0)
        seq_buffer << voice_line[counter % voice_line.size()]
      end
    end
    pattern << seq_buffer
    counter += 1
  end
  pattern = separate(pattern, step_size)
  pattern
end


pattern = make_pattern("0 - 0", 0.5)
puts pattern
